app/src/main/java/ch/epfl/sdp/appart/database/FirestoreDatabaseService.java

```diff
@@ -70,32 +71,54 @@ public FirestoreDatabaseService() {
                 task -> {
-
-                    List<Card> queriedCards = new ArrayList<>();
-
-                    if (task.isSuccessful()) {
-
-                        for (QueryDocumentSnapshot document : task.getResult()) {
-                            Map<String, Object> data = document.getData();
-                            queriedCards.add(
-                                    new Card(document.getId(), (String) data.get(CardLayout.AD_ID), (String) data.get(CardLayout.USER_ID),
-                                            (String) data.get(CardLayout.CITY),
-                                            (long) data.get(CardLayout.PRICE),
-                                            (String) data.get(CardLayout.IMAGE)));
-                        }
-                        result.complete(queriedCards);
-
-                    } else {
-                        result.completeExceptionally(
-                                new DatabaseServiceException(
-                                        "failed to fetch the cards from firebase"
-                                ));
-                    }
+                    handleTask(task,  result);
                 }
         );
+        return result;
+    }
+
+    @NotNull
+    @Override
+    @NonNull
+    public CompletableFuture<List<Card>> getCardsFilter(String location) {
+        CompletableFuture<List<Card>> result = new CompletableFuture<>();
+        //ask firebase async to get the cards objects and notify the future
+        //when they have been fetched
+        db.collection(FirebaseLayout.CARDS_DIRECTORY)
+            .whereGreaterThanOrEqualTo("city",  location)
+            .whereLessThanOrEqualTo("city", location+"\uF7FF").get().addOnCompleteListener(
+            task -> {
+                handleTask(task,  result);
+            }
+        );
 
         return result;
     }
+    private Card generateCard(Map<String, Object> data,QueryDocumentSnapshot document){
+        return new Card(
+            document.getId(),
+            (String) data.get(CardLayout.AD_ID),
+            (String) data.get(CardLayout.USER_ID),
+            (String) data.get(CardLayout.CITY),
+            (long) data.get(CardLayout.PRICE),
+            (String) data.get(CardLayout.IMAGE));
+    }
+
+    private void handleTask(Task task, CompletableFuture<List<Card>> result){
+        List<Card> queriedCards = new ArrayList<>();
+        if (task.isSuccessful()) {
+            for (QueryDocumentSnapshot document : (Iterable<? extends QueryDocumentSnapshot>) task.getResult()) {
+                Map<String, Object> data = document.getData();
+                queriedCards.add(generateCard(data, document));
+            }
+            result.complete(queriedCards);
+        } else {
+            result.completeExceptionally(
+                new DatabaseServiceException(
+                    "failed to fetch the cards from firebase"
```

Comment:
Maybe getting the message from the task (task.getException().getMessage()) would give more useful hints as to what caused the failure.

Reply:
okay, I added to the exception message.